Paintroid/src/test/java/org/catrobat/paintroid/test/presenter/MainActivityPresenterTest.java

```diff
@@ -1007,6 +955,132 @@ public void testOnSaveImagePostExecuteWhenFailedThenShowDialog() {
 		verify(navigator).showSaveErrorDialog();
 	}
 
+	@Test
+	public void testSaveImageClickedWhenPermissionNotGrantedThenAskForPermission() {
+		doReturn(false).when(navigator).doIHavePermission(Manifest.permission.WRITE_EXTERNAL_STORAGE);
+		doReturn(true).when(navigator).isSdkAboveOrEqualM();
+
+		presenter.saveImageClicked();
+
+		verify(navigator).askForPermission(new String[] {Manifest.permission.WRITE_EXTERNAL_STORAGE},
+				PERMISSION_EXTERNAL_STORAGE_SAVE);
+	}
+
+	@Test
+	public void testHandlePermissionRequestResultForRequestAfterSaveClickedPermissionNotGranted() {
+		presenter.handlePermissionRequestResults(PERMISSION_EXTERNAL_STORAGE_SAVE,
+				new String[] {Manifest.permission.WRITE_EXTERNAL_STORAGE},
+				new int[] {});
+
+		verify(navigator).showPermissionDialog(PermissionInfoDialog.PermissionType.EXTERNAL_STORAGE,
+				EXTERNAL_STORAGE_PERMISSION_DIALOG,
+				PERMISSION_EXTERNAL_STORAGE_SAVE);
+	}
+
+	@Test
+	public void testHandlePermissionRequestResultForRequestAfterSaveClickedPermissionGranted() {
+		presenter.handlePermissionRequestResults(PERMISSION_EXTERNAL_STORAGE_SAVE,
+				new String[] {Manifest.permission.WRITE_EXTERNAL_STORAGE},
+				new int[] {PackageManager.PERMISSION_GRANTED});
+
+		verify(presenter).saveImageClicked();
+	}
+
+	@Test
+	public void testSaveCopyClickedWhenPermissionNotGrantedThenAskForPermission() {
+		doReturn(false).when(navigator).doIHavePermission(Manifest.permission.WRITE_EXTERNAL_STORAGE);
+		doReturn(true).when(navigator).isSdkAboveOrEqualM();
+
+		presenter.saveCopyClicked();
+
+		verify(navigator).askForPermission(new String[] {Manifest.permission.WRITE_EXTERNAL_STORAGE},
+				PERMISSION_EXTERNAL_STORAGE_SAVE_COPY);
+	}
+
+	@Test
+	public void testHandlePermissionRequestResultForRequestAfterSaveCopyClickedPermissionNotGranted() {
+		presenter.handlePermissionRequestResults(PERMISSION_EXTERNAL_STORAGE_SAVE_COPY,
+				new String[] {Manifest.permission.WRITE_EXTERNAL_STORAGE},
+				new int[] {});
+
+		verify(navigator).showPermissionDialog(PermissionInfoDialog.PermissionType.EXTERNAL_STORAGE,
+				EXTERNAL_STORAGE_PERMISSION_DIALOG,
+				PERMISSION_EXTERNAL_STORAGE_SAVE_COPY);
+	}
+
+	@Test
+	public void testHandlePermissionRequestResultForRequestAfterSaveCopyClickedPermissionGranted() {
+		presenter.handlePermissionRequestResults(PERMISSION_EXTERNAL_STORAGE_SAVE_COPY,
+				new String[] {Manifest.permission.WRITE_EXTERNAL_STORAGE},
+				new int[] {PackageManager.PERMISSION_GRANTED});
+
+		verify(presenter).saveCopyClicked();
```

Comment:
Regarding `@spy` - I see now.
Please don't use mockito to verify that the presenter called an internal function.
You should check that the presenter did the right things to the collaborating classes, in this case you should verify that the presenter calls `interactor.saveCopy` does _not_ call `navigator.askForPermission()` again.
Don't mock/spy the class under test

Reply:
Okay